llm/openai/openai.go

```diff
@@ -219,27 +253,29 @@ func (o *OpenAI) stream(
 		return fmt.Errorf("%w: %w", ErrOpenAIChat, err)
 	}
 
-	var messages []*thread.Message
 	var content string
+	var messages []*thread.Message
+	var allToolCalls []openai.ToolCall
+	var currentToolCall openai.ToolCall
 	for {
 		response, errRecv := stream.Recv()
 		if errors.Is(errRecv, io.EOF) {
-			o.streamCallbackFn(EOS)
-
-			if len(content) > 0 {
-				messages = append(messages, thread.NewAssistantMessage().AddContent(
-					thread.NewTextContent(content),
-				))
-			}
+			messages = o.processErrorResponse(messages, content, &currentToolCall, allToolCalls)
 			break
 		}
 
 		if len(response.Choices) == 0 {
 			return fmt.Errorf("%w: no choices returned", ErrOpenAIChat)
 		}
 
-		if response.Choices[0].FinishReason == "tool_calls" || len(response.Choices[0].Delta.ToolCalls) > 0 {
-			messages = append(messages, o.callTools(response.Choices[0].Delta.ToolCalls)...)
+		if isStreamToolCallResponse(&response) {
+			updatedToolCall, isNewTool := consumeToolCallResponse(&response, &currentToolCall)
```

Comment:
Use the new function name.

```suggestion
			updatedToolCall, isNewTool := handleStreamToolCallResponse(&response, &currentToolCall)
```